Record tensor factory functions in trace

zdevito opened the pull request:
Things like torch.zeros now appear in traces rather than constants.

To continue to support our current level of ONNX export, we run
constant prop to turn these back into constants where possible before
export.

zdevito committed:
Record tensor factory functions in trace

Things like torch.zeros now appear in traces rather than constants.

To continue to support our current level of ONNX export, we run
constant prop to turn these back into constants where possible before
export.

zdevito committed:
Appease compilers

zdevito committed:
TORCH_API

apaszke reviewed (commented):
Mostly LGTM, but we should make sure we don't embed large short-lived tensors as constants into our traces.

zdevito commented on test/test_jit.py:
```diff
@@ -4685,8 +4701,9 @@ def __init__(self, m):
                 x += x
-                if True:
-                    if True:
+                c = sum(x) > 4  # can't be True, onnx export constant props
```
because constant prop will just remove the if statements and this test is testing that if statements export correctly, I will clarify the comment.

zdevito commented on tools/autograd/gen_variable_type.py:
```diff
@@ -183,6 +183,41 @@ def should_trace(declaration):
+
+    local['trace_outputs'] = get_trace_outputs(declaration)
+
+    return (PRE_RECORD_TRACE.substitute(local), POST_RECORD_TRACE.substitute(local))
```
This is a move. Minor changes (e.g. `arguments -> declaration['arguments']`) so it works as a global function.

zdevito commented on torch/csrc/jit/python_ir.cpp:
```diff
@@ -433,14 +433,27 @@ void initPythonIRBindings(PyObject * module_) {
+        case TypeKind::FloatType:
+          return "FloatType";
+        case TypeKind::StringType:
+          return "StringType";
```
These are used as enum tags in Python (e.g. `foo.type() == "StringType"`), it wouldn't be a good idea to use pretty printing. I edited it so that we will get a warning if you add a enum case.

zdevito replied:
sorry (`foo.type().kind() == "StringType"`)

zdevito commented on torch/onnx/utils.py:
```diff
@@ -103,24 +104,32 @@ def export(model, args, f, export_params=True, verbose=False, training=False,
+    # into a trace where we previously recorded constants
+    # use constant prop to maintain our current level of onnx support
+    # without implementing symbolics for all of them
+    torch._C._jit_pass_constant_propagation(graph)
```
Offline discussion: we should limit how much we constant prop through factories in non-ONNX code. However, this patch doesn't regress that behavior in anyway, so I am going to defer it a later task.

zdevito committed:
PR Comments

zdevito committed:
Fix constant prop bug

Constant prop needs to work on prim::ListConstruct even when the list
has 0 elements.

zdevito committed:
appease windows

zdevito committed:
Rebase fix

zdevito committed:
Fix more constant prop bugs

zdevito committed:
ONNX sometimes pre-inserts ops into the graph,
make sure constant prop doesnt try to execute them

jamesr66a reviewed (approved):
ONNX changes LGTM
